### ISSUE TITLE
Freeze while starting chrome for system tests 

### ISSUE THREAD

**Issue by feerrenrut** [nvaccess/nvda#11339]
### Steps to reproduce:
Intermittent system test failure:
See build https://ci.appveyor.com/project/NVAccess/nvda/builds/33933406/artifacts

[ChromeFreeze-SystemTestResult.zip](https://github.com/nvaccess/nvda/files/4879960/ChromeFreeze-SystemTestResult.zip)

### Actual behavior:
When gaining focus Chrome / NVDA seemed to freeze and the watchdog was triggered giving the following errors:
```
DEBUGWARNING - watchdog._watcher (14:39:27.712) - watchdog (5184):
Trying to recover from freeze, core stack:
  File "nvda.pyw", line 215, in <module>
  File "core.pyc", line 545, in main
  File "wx\core.pyc", line 2134, in MainLoop
  File "gui\__init__.pyc", line 1049, in Notify
  File "core.pyc", line 514, in run
  File "IAccessibleHandler\__init__.pyc", line 903, in pumpAll
  File "IAccessibleHandler\__init__.pyc", line 578, in processGenericWinEvent
  File "IAccessibleHandler\__init__.pyc", line 539, in winEventToNVDAEvent
  File "NVDAObjects\IAccessible\__init__.pyc", line 48, in getNVDAObjectFromEvent
  File "NVDAObjects\__init__.pyc", line 79, in __call__
  File "NVDAObjects\IAccessible\__init__.pyc", line 605, in __init__
  File "baseObject.pyc", line 25, in __get__
  File "NVDAObjects\IAccessible\__init__.pyc", line 1619, in _get_IA2WindowHandle
  File "comtypesMonkeyPatches.pyc", line 26, in __call__

DEBUGWARNING - core.CorePump.run (14:39:27.872) - MainThread (904):
errors in this core pump cycle
Traceback (most recent call last):
  File "core.pyc", line 514, in run
  File "IAccessibleHandler\__init__.pyc", line 903, in pumpAll
  File "IAccessibleHandler\__init__.pyc", line 578, in processGenericWinEvent
  File "IAccessibleHandler\__init__.pyc", line 539, in winEventToNVDAEvent
  File "NVDAObjects\IAccessible\__init__.pyc", line 48, in getNVDAObjectFromEvent
  File "NVDAObjects\__init__.pyc", line 79, in __call__
  File "NVDAObjects\IAccessible\__init__.pyc", line 605, in __init__
  File "baseObject.pyc", line 25, in __get__
  File "NVDAObjects\IAccessible\__init__.pyc", line 1619, in _get_IA2WindowHandle
  File "comtypesMonkeyPatches.pyc", line 34, in __call__
core.CallCancelled: COM call cancelled
DEBUGWARNING - NVDAObjects.IAccessible.IAccessible._get_IAccessibleRole (14:39:28.393) - MainThread (904):
accRole failed: (-2147024809, 'The parameter is incorrect.', (None, None, None, 0, None))
DEBUGWARNING - watchdog._watcher (14:39:28.409) - watchdog (5184):
Trying to recover from freeze, core stack:
  File "nvda.pyw", line 215, in <module>
  File "core.pyc", line 545, in main
  File "wx\core.pyc", line 2134, in MainLoop
  File "gui\__init__.pyc", line 1049, in Notify
  File "core.pyc", line 514, in run
  File "IAccessibleHandler\__init__.pyc", line 903, in pumpAll
  File "IAccessibleHandler\__init__.pyc", line 578, in processGenericWinEvent
  File "IAccessibleHandler\__init__.pyc", line 539, in winEventToNVDAEvent
  File "NVDAObjects\IAccessible\__init__.pyc", line 48, in getNVDAObjectFromEvent
  File "NVDAObjects\__init__.pyc", line 69, in __call__
  File "NVDAObjects\__init__.pyc", line 224, in findBestAPIClass
  File "NVDAObjects\IAccessible\MSHTML.pyc", line 464, in kwargsFromSuper
  File "NVDAObjects\IAccessible\MSHTML.pyc", line 180, in HTMLNodeFromIAccessible
  File "comtypes\__init__.pyc", line 1191, in QueryService
  File "comtypesMonkeyPatches.pyc", line 26, in __call__

DEBUGWARNING - core.CorePump.run (14:39:28.425) - MainThread (904):
errors in this core pump cycle
Traceback (most recent call last):
  File "core.pyc", line 514, in run
  File "IAccessibleHandler\__init__.pyc", line 903, in pumpAll
  File "IAccessibleHandler\__init__.pyc", line 578, in processGenericWinEvent
  File "IAccessibleHandler\__init__.pyc", line 539, in winEventToNVDAEvent
  File "NVDAObjects\IAccessible\__init__.pyc", line 48, in getNVDAObjectFromEvent
  File "NVDAObjects\__init__.pyc", line 86, in __call__
  File "NVDAObjects\IAccessible\__init__.pyc", line 544, in findOverlayClasses
  File "NVDAObjects\IAccessible\chromium.pyc", line 61, in findExtraOverlayClasses
  File "NVDAObjects\IAccessible\ia2Web.pyc", line 207, in findExtraOverlayClasses
  File "baseObject.pyc", line 41, in __get__
  File "baseObject.pyc", line 144, in _getPropertyViaCache
  File "NVDAObjects\IAccessible\__init__.pyc", line 1094, in _get_IA2Attributes
  File "comtypesMonkeyPatches.pyc", line 34, in __call__
core.CallCancelled: COM call cancelled
```

From the log, it looks like when the window loses then gains focus (due to the quit dialog), it becomes unblocked.

### Expected behavior:
No freeze on startup of chrome.

### System configuration
Running NVDA on appveyor with chrome.

#### NVDA installed/portable/running from source:
Installed

#### NVDA version:
2020.3 in dev.

#### Windows version:
Windows 10
#### Name and version of other software in use when reproducing the issue:
Chrome, unsure of version.

**Comment by feerrenrut** [comment]
It would be useful for situations like this (and others) to be able to enable a minidump collection of the focused process when the watchdog notices a freeze.

**Comment by feerrenrut** [comment]
Another example: https://ci.appveyor.com/project/NVAccess/nvda/builds/34111456/artifacts

**Comment by feerrenrut** [comment]
Now that we get stack traces from all threads, a new example:
https://ci.appveyor.com/project/NVAccess/nvda/builds/34566006/artifacts

@michaelDCurran this seems to show the watchdog trying to recover from the freeze after only ~2 seconds, I thought the limit was higher than this. It would be nice if this message included the amount of time the watchdog has tracked since the last "still alive" call. 

Either way, I think this does help to point out an issue with the way the system tests are working, while waiting for some speech, this will sleep in a loop. Perhaps the Robot remote server should be moved to another thread. So the blocking call can't hold up NVDA's core loop.
```
IO - speech.speak (01:31:45.707) - MainThread (5392):
Speaking [LangChangeCommand ('en'), 'Recycle Bin', 'not selected', '1 of 7']
DEBUGWARNING - watchdog._watcher (01:31:47.542) - watchdog (3836):
Trying to recover from freeze. Listing stacks for Python threads:
Python stack for thread 3836 (watchdog):
  File "threading.pyc", line 890, in _bootstrap
  File "threading.pyc", line 926, in _bootstrap_inner
  File "threading.pyc", line 870, in run
  File "watchdog.pyc", line 112, in _watcher
  File "watchdog.pyc", line 62, in getFormattedStacksForAllThreads

Python stack for thread 3224 (Thread-1):
  File "threading.pyc", line 890, in _bootstrap
  File "threading.pyc", line 926, in _bootstrap_inner
  File "threading.pyc", line 870, in run
  File "C:\Users\appveyor\AppData\Local\Temp\1\nvdaProfile\scratchpad\globalPlugins\speechSpyGlobalPlugin\libs\robotremoteserver.py", line 135, in serve
    self._server.serve()
  File "C:\Users\appveyor\AppData\Local\Temp\1\nvdaProfile\scratchpad\globalPlugins\speechSpyGlobalPlugin\libs\robotremoteserver.py", line 213, in serve
    self.serve_forever()
  File "socketserver.pyc", line 237, in serve_forever
  File "socketserver.pyc", line 316, in _handle_request_noblock
  File "socketserver.pyc", line 347, in process_request
  File "socketserver.pyc", line 360, in finish_request
  File "socketserver.pyc", line 720, in __init__
  File "http\server.pyc", line 426, in handle
  File "http\server.pyc", line 414, in handle_one_request
  File "C:\Users\appveyor\AppData\Local\Temp\1\nvdaProfile\scratchpad\globalPlugins\speechSpyGlobalPlugin\libs\xmlrpc\server.py", line 516, in do_POST
    data, getattr(self, '_dispatch', None), self.path
  File "C:\Users\appveyor\AppData\Local\Temp\1\nvdaProfile\scratchpad\globalPlugins\speechSpyGlobalPlugin\libs\xmlrpc\server.py", line 263, in _marshaled_dispatch
    response = self._dispatch(method, params)
  File "C:\Users\appveyor\AppData\Local\Temp\1\nvdaProfile\scratchpad\globalPlugins\speechSpyGlobalPlugin\libs\xmlrpc\server.py", line 411, in _dispatch
    return func(*params)
  File "C:\Users\appveyor\AppData\Local\Temp\1\nvdaProfile\scratchpad\globalPlugins\speechSpyGlobalPlugin\libs\robotremoteserver.py", line 175, in run_keyword
    return self._library.run_keyword(name, args, kwargs)
  File "C:\Users\appveyor\AppData\Local\Temp\1\nvdaProfile\scratchpad\globalPlugins\speechSpyGlobalPlugin\libs\robotremoteserver.py", line 299, in run_keyword
    return KeywordRunner(kw).run_keyword(args, kwargs)
  File "C:\Users\appveyor\AppData\Local\Temp\1\nvdaProfile\scratchpad\globalPlugins\speechSpyGlobalPlugin\libs\robotremoteserver.py", line 409, in run_keyword
    return_value = self._keyword(*args, **kwargs)
  File "C:\Users\appveyor\AppData\Local\Temp\1\nvdaProfile\scratchpad\globalPlugins\speechSpyGlobalPlugin\__init__.py", line 203, in wait_for_specific_speech
    errorMessage=None
  File "C:\Users\appveyor\AppData\Local\Temp\1\nvdaProfile\scratchpad\globalPlugins\speechSpyGlobalPlugin\blockUntilConditionMet.py", line 46, in _blockUntilConditionMet
    _sleep(SLEEP_TIME)

Python stack for thread 2728 (winInputHook):
  File "threading.pyc", line 890, in _bootstrap
  File "threading.pyc", line 926, in _bootstrap_inner
  File "threading.pyc", line 870, in run
  File "winInputHook.pyc", line 79, in hookThreadFunc

Python stack for thread 3908 (_UIAHandler.UIAHandler.MTAThread):
  File "threading.pyc", line 890, in _bootstrap
  File "threading.pyc", line 926, in _bootstrap_inner
  File "threading.pyc", line 870, in run
  File "_UIAHandler.pyc", line 310, in MTAThreadFunc
  File "queue.pyc", line 170, in get
  File "threading.pyc", line 296, in wait

Python stack for thread 3888 (braille._BgThread):
  File "threading.pyc", line 890, in _bootstrap
  File "threading.pyc", line 926, in _bootstrap_inner
  File "threading.pyc", line 870, in run
  File "braille.pyc", line 2187, in func

Python stack for thread 5392 (MainThread):
  File "nvda.pyw", line 215, in <module>
  File "core.pyc", line 545, in main
  File "wx\core.pyc", line 2134, in MainLoop
  File "gui\__init__.pyc", line 1050, in Notify
  File "core.pyc", line 514, in run
  File "IAccessibleHandler\__init__.pyc", line 907, in pumpAll
  File "IAccessibleHandler\__init__.pyc", line 582, in processGenericWinEvent
  File "IAccessibleHandler\__init__.pyc", line 543, in winEventToNVDAEvent
  File "NVDAObjects\IAccessible\__init__.pyc", line 42, in getNVDAObjectFromEvent
  File "IAccessibleHandler\__init__.pyc", line 345, in accessibleObjectFromEvent
  File "oleacc.pyc", line 265, in AccessibleObjectFromEvent
```

**Comment by michaelDCurran** [comment]
RobotRemoteServer is shown in the second Python stack, which is not the 
core loop. The core loop (main thread) is the very last one, which seems 
to be stuck in a call to AccessibleObjectFromEvent.


Watchdog looks for freezes much more rapidly any time there is a 
foreground change. I think it might be 0.5 seconds?

@jcsteh might be able to better explain the ideas behind why we do this 
more rapidly on foreground changes. But essentially the idea is that if 
the foreground changes, then NVDA should be always responsive. If it was 
stuck somehow in the old app, it should break out to handle the new app. 
E.g. a user hears nothing from NvDA then presses alt+tab.

In general this does not cause us harm, but if this is always an issue 
with the system tests, then you could consider disabling or suspending 
watchdog while Chrome is started.

We could also consider changing watchdog so that it only goes into fast 
recovery mode if the foreground changes and the user provides input 
(presses a key etc).

**Comment by feerrenrut** [comment]
I believe this issue is now resolved. There has been a number of changes that will at the very least impact the way this presents. If this is encountered again a new issue should be created.
